src/bin/deku_cli.ml

```diff
@@ -220,45 +218,46 @@ let create_transaction node_folder sender_wallet_file received_address amount
         validator_uris in
     match validator_uris with
     | [] -> Lwt.return (`Error (false, "No validators found"))
-    | (_, validator_uri) :: _ ->
-      let%await block_level_response = request_block_level () validator_uri in
-      let block_level = block_level_response.level in
-      let%await wallet = Files.Wallet.read ~file:sender_wallet_file in
-      let operation =
-        match (Address.to_key_hash received_address, argument) with
-        | Some addr, None ->
-          Core_deku.User_operation.make ~source:wallet.address
-            (Transaction { destination = addr; amount; ticket })
-        | Some _, Some _ ->
-          failwith "can't pass an argument to implicit account"
-        | None, None -> failwith "Invalid transaction"
-        | None, Some arg ->
-          let payload =
-            match vm_flavor with
-            | `Lambda -> Contract_vm.Invocation_payload.lambda_of_yojson ~arg
-            | `Dummy -> Contract_vm.Invocation_payload.dummy_of_yojson ~arg
-          in
-          let arg = payload |> Result.get_ok in
-          Core_deku.User_operation.make ~source:wallet.address
-            (Contract_invocation
-               {
-                 to_invoke =
-                   Address.to_contract_hash received_address |> Option.get;
-                 argument = arg;
-               }) in
-      let transaction =
-        Protocol.Operation.Core_user.sign ~secret:wallet.priv_key
-          ~nonce:(Crypto.Random.int32 Int32.max_int)
-          ~block_height:block_level ~data:operation in
+    | validator_uri :: _ -> f validator_uri)
 
-      let%await identity = read_identity ~node_folder in
-      let%await () =
-        Network.request_user_operation_gossip
-          { user_operation = transaction }
-          identity.uri in
-      Format.printf "operation.hash: %s\n%!"
-        (BLAKE2B.to_string transaction.hash);
-      Lwt.return (`Ok ()))
+let create_transaction node_folder sender_wallet_file received_address amount
+    ticket argument vm_flavor =
+  let open Network in
+  with_validator_uri node_folder @@ fun (_, validator_uri) ->
+  let%await block_level_response = request_block_level () validator_uri in
+  let block_level = block_level_response.level in
+  let%await wallet = Files.Wallet.read ~file:sender_wallet_file in
+  let operation =
+    match (Address.to_key_hash received_address, argument) with
```

Comment:
I guess `create_transaction` should be splitted into two different functions.
We only do things if `received_address` is an Implicit address and `argument` is None.
And if `argument` is `Some argument` and `received_address` is None...
I don't know if I am clear enough, but I think we are trying to handle two different business cases in the same function.

Reply:
Sorry, I gave a very poor PR description. I updated it to make it clear exactly what I intend to refactor in this PR. That said, your proposal is outside my intended scope, but it's probably a good one to do. PR's accepted :smile: